gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -536,48 +565,46 @@ private boolean isByteBufferBeyondCurrentRequestRange(ByteBuffer byteBuffer) {
 
   private int readObjectContentFromGCS(ByteBuffer byteBuffer) throws IOException {
     int bytesRead = 0;
-    while (byteBuffer.hasRemaining() && moreServerContent()) {
-      ReadObjectResponse res = resIterator.next();
+    ReadObjectResponse res = resIterator.next();
 
-      // When zero-copy mashaller is used, the stream that backs GetObjectMediaResponse
-      // should be closed when the mssage is no longed needed so that all buffers in the
-      // stream can be reclaimed. If zero-copy is not used, stream will be null.
-      InputStream stream = getObjectMediaResponseMarshaller.popStream(res);
-      try {
-        ByteString content = res.getChecksummedData().getContent();
-        if (bytesToSkipBeforeReading >= 0 && bytesToSkipBeforeReading < content.size()) {
-          content = res.getChecksummedData().getContent().substring((int) bytesToSkipBeforeReading);
-          positionInGrpcStream += bytesToSkipBeforeReading;
-          bytesToSkipBeforeReading = 0;
-        } else if (bytesToSkipBeforeReading >= content.size()) {
-          positionInGrpcStream += content.size();
-          bytesToSkipBeforeReading -= content.size();
-          continue;
-        }
+    // When zero-copy marshaller is used, the stream that backs GetObjectMediaResponse
+    // should be closed when the message is no longed needed so that all buffers in the
+    // stream can be reclaimed. If zero-copy is not used, stream will be null.
+    InputStream stream = getObjectMediaResponseMarshaller.popStream(res);
+    try {
+      ByteString content = res.getChecksummedData().getContent();
+      if (bytesToSkipBeforeReading >= 0 && bytesToSkipBeforeReading < content.size()) {
+        content = res.getChecksummedData().getContent().substring((int) bytesToSkipBeforeReading);
```

Comment:
Is there a risk of calling `res.getChecksummedData()` twice above? Either with respect to idempotency or latency. Wouldn't this be better?
```
if ......
  content = content.substring((int) bytesToSkipBeforeReading);
...
```

Reply:
can we check if substring is just pointer manipulation and not actual data copy.

Reply:
Done.

Yes, [Documentation](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/ByteString.html#substring-int-int-) suggests it not actual data copy